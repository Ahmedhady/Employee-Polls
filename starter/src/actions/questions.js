import { saveQuestion, saveQuestionAnswer } from "../utils/api"; 
import { showLoading, hideLoading } from "react-redux-loading-bar";
import { addQuestionUser, addAnswerUser } from "./users";
import { _saveQuestionAnswer, _saveQuestion } from "../utils/_DATA";

export const RECEIVE_QUESTIONS = 'RECEIVE_QUESTIONS'
export const ADD_QUESTION = 'ADD_QUESTION';
export const ADD_ANSWER_QUESTION = 'ADD_ANSWER_QUESTION';

export function receiveQuestions(questions) {
    return {
        type: RECEIVE_QUESTIONS,
        questions,
    }
}

function addQuestion(question) {
    return {
        type: ADD_QUESTION,
        question,
    }
}

export function addAnswerQuestion(author, qid, answer) {
    return {
        type: ADD_ANSWER_QUESTION,
        author,
        qid,
        answer,
    };
}

export function handleAddQuestion(optionOneText, optionTwoText){
    return (dispatch, getState) => {
        const { authedUser } = getState();

        dispatch (showLoading());

        return saveQuestion(optionOneText, optionTwoText, authedUser)
        .then((question) => {
            dispatch(addQuestion(question))
            dispatch(addQuestionUser(question))
        })
        .then(() => dispatch(hideLoading()));
    }
}

/*export function handleAnswerQuestion(questionId, answer){
    return (dispatch, getState) => {
        const { authedUser } = getState();

        //dispatch(showLoading());

        return saveQuestionAnswer(authedUser.id, questionId, answer)
        .then(() => {
            dispatch(addAnswerQuestion(authedUser.id, questionId, answer))
            dispatch(addAnswerUser(authedUser.id, questionId, answer))
        })
        //.then(() => dispatch(hideLoading()));
    }
}*/

export function handleAddAnswer(questionId, answer) {
    return (dispatch, getState) => {
        const { authedUser } = getState();
        dispatch(showLoading());
        return saveQuestionAnswer(authedUser, questionId, answer)
            .then(() => {
                dispatch(addAnswerQuestion(authedUser, questionId, answer));
                dispatch(addAnswerUser(authedUser, questionId, answer));
            })
            .then(() => dispatch(hideLoading()));
    };
}

/*export function handleAddAnswer(qid, answer) {
    return (dispatch, getState) => {
      const  { authedUser } = getState();
      dispatch(showLoading());
  
      return saveQuestionAnswer({
        authedUser,
        qid,
        answer
      })
        .then(() => { 
          dispatch(addAnswerQuestion({qid, answer, authedUser}))
          dispatch(addAnswerUser({qid, answer, authedUser}))
        })  
        .then(() => dispatch(hideLoading()));
    }
  }*/
