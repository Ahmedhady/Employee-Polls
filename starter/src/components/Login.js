import { connect } from "react-redux";
import { setAuthedUser } from "../actions/authedUser";
import { useState } from "react";
import { useNavigate, useLocation } from "react-router-dom";
import LOGO from '../assets/logo.jpeg';

const Login = (props) => {

    const navigate = useNavigate();
    const state = useLocation();
  
    const [userName, setUserName] = useState("");
    const [password, setUserPassword] = useState("");
    const [errorMessage, setErrorMessage] = useState("");
  
    const autheduser = Object.values(props.users).filter((user) => user.id === userName && user.password === password);
  
    const handleLogIn = (event) => {
      event.preventDefault();
      if (!autheduser || userName === "" || password === "" || autheduser.length === 0) {
        setErrorMessage("Please try to login again.");
      } else {props.dispatch(setAuthedUser(autheduser[0].id));
        navigate(state?.path || "/");
      }
    };
  
    const handleUserName = (event) => {
        setUserName(event.target.value);
    };
  
    const handleUserPassword = (event) => {
        setUserPassword(event.target.value);
    };

    const isEmpty = () => userName === "" || password === "";
  
    return (
      <p className="center">
        <h1>Employee Polls</h1>
        <div><img src={LOGO} alt="logo" className="avatar" /></div>
        <div>
        Log In
        <form onSubmit={handleLogIn}>
            <p>
          <input type="text" placeholder={"id"} onChange={handleUserName}/>
          </p>
          <p>
          <input type="password" placeholder={"password"} onChange={handleUserPassword}/>
          </p>
          <p>
          <button className="btn" disabled={isEmpty()}>Login</button>
          </p>
          <p>
            {errorMessage}
          </p>
        </form>
        </div>
      </p>
    );
  };
  
  const mapStateToProps = ( props ) => (
    props
)
  
  export default connect(mapStateToProps)(Login);