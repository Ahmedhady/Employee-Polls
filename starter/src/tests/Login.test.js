import Login from "../components/Login";
import { render, fireEvent } from "@testing-library/react";
import * as React from "react";
import { Provider } from "react-redux";
import { MemoryRouter } from "react-router";
import reducer from "../reducers";
import middleware from "../middleware";
import thunk from 'redux-thunk';
import { createStore } from "redux";
import { applyMiddleware } from "redux";
import { BrowserRouter as Router } from "react-router-dom";
import { handleInitialData } from "../actions/shared";

const store = createStore(reducer, applyMiddleware(thunk));
//const store = createStore(reducer, middleware);

describe("login", () => {
  it("matches the snapshot", () => {
    const component = render(<MemoryRouter><Provider store={store}><Login /></Provider></MemoryRouter>);
    expect(component).toMatchSnapshot();
  });

  it("verify of Enter name & password and login", () => {
    const component = render(<MemoryRouter><Provider store={store}><Login /></Provider></MemoryRouter>);
    expect(component.getByTestId("testId-name-input")).toBeInTheDocument();
    expect(component.getByTestId("testId-password-input")).toBeInTheDocument();
    expect(component.getByTestId("testId-submit-button")).toBeInTheDocument();
    });

  it("verify of no error with valid name and password", () => {
    const component = render(<MemoryRouter><Provider store={store}><Login /></Provider></MemoryRouter>);

    const nameInput = component.getByTestId("testId-name-input");
    fireEvent.change(nameInput, { target: { value: "tylermcginnis" } });
    expect(nameInput).toBeInTheDocument();

    const passwordInput = component.getByTestId("testId-password-input");
    fireEvent.change(passwordInput, { target: { value: "abc321" } });
    expect(passwordInput).toBeInTheDocument();

    const submitButton = component.getByTestId("testId-submit-button");
    fireEvent.click(submitButton);
    expect(submitButton).toBeInTheDocument();
  });
});