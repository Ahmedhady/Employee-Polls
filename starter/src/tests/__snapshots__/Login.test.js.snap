// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`login matches the snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <p
        class="center"
      >
        <h1>
          Employee Polls
        </h1>
        <div>
          <img
            alt="logo"
            class="avatar"
            src="logo.jpeg"
          />
        </div>
        <div>
          <form>
            <p>
              <input
                data-testid="testId-name-input"
                placeholder="id"
                type="text"
              />
            </p>
            <p>
              <input
                data-testid="testId-password-input"
                placeholder="password"
                type="password"
              />
            </p>
            <p>
              <button
                class="btn"
                data-testid="testId-submit-button"
                disabled=""
              >
                Login
              </button>
            </p>
            <p
              data-testid="testId-error"
            />
          </form>
        </div>
      </p>
    </div>
  </body>,
  "container": <div>
    <p
      class="center"
    >
      <h1>
        Employee Polls
      </h1>
      <div>
        <img
          alt="logo"
          class="avatar"
          src="logo.jpeg"
        />
      </div>
      <div>
        <form>
          <p>
            <input
              data-testid="testId-name-input"
              placeholder="id"
              type="text"
            />
          </p>
          <p>
            <input
              data-testid="testId-password-input"
              placeholder="password"
              type="password"
            />
          </p>
          <p>
            <button
              class="btn"
              data-testid="testId-submit-button"
              disabled=""
            >
              Login
            </button>
          </p>
          <p
            data-testid="testId-error"
          />
        </form>
      </div>
    </p>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
